### PR TITLE
Loki crds

### LINKED ISSUE #3395: Grafana Agent Operator - Support creating rules for Loki
Similar to #1986 and #2604 there is now support to configure Mimir's ruler with rules from the prom-operator CRDs. Could we use the same CRDs to also generate Loki rules as well?

### PR DESCRIPTION
**Pull request by EStork09**
#### Which issue(s) this PR fixes

Fixes #3395 

#### Notes to the Reviewer
Copied #2604 basically.
Would remove the 404 skip check when [#9489](https://github.com/grafana/loki/issues/9489) closes on the loki repo.
Had to use `-` to seperate rule groups/namespaces, `/` was not properly passing through the routes, it appears loki needs to align with mimir to handle those, loki is just using [Path](https://github.com/grafana/loki/blob/main/pkg/loki/modules.go#L980) while mimir is using [PathPrefix](https://github.com/grafana/mimir/blob/main/pkg/api/api.go#L303).

#### PR Checklist

- [x] CHANGELOG updated
- [x] Documentation added
- [x] Tests updated

### CODE REVIEW

**Review by tpaschalis — COMMENTED** [review]
Thanks for the PR!

This LGTM at a first glance with a few nits, but I haven't tested this yet and would like to before we check it in.

File: docs/sources/flow/reference/components/loki.rules.kubernetes.md [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+
+# loki.rules.kubernetes
+
+{{< docs/shared lookup="flow/stability/beta.md" source="agent" >}}
+
+`loki.rules.kubernetes` discovers `PrometheusRule` Kubernetes resources and
+loads them into a Loki instance.
+
+* Multiple `loki.rules.kubernetes` components can be specified by giving them
+  different labels.
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/loki.rules.kubernetes.md:15]:
```suggestion
* You can specify multiple `loki.rules.kubernetes` components by giving them different labels.
```

File: docs/sources/flow/reference/components/loki.rules.kubernetes.md [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+* [Kubernetes label selectors][] can be used to limit the `Namespace` and
+  `PrometheusRule` resources considered during reconciliation.
+* Compatible with the Ruler APIs of Grafana Loki, Grafana Cloud, and Grafana Enterprise Metrics.
+* Compatible with the `PrometheusRule` CRD from the [prometheus-operator][].
+* This component accesses the Kubernetes REST API from [within a Pod][].
+
+> **NOTE**: This component requires [Role-based access control (RBAC)][] to be setup
+> in Kubernetes in order for the Agent to access it via the Kubernetes REST API.
+> For an example RBAC configuration please click [here](#example).
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/loki.rules.kubernetes.md:24]:
```suggestion
{{% admonition type="note" %}}
This component requires [Role-based access control (RBAC)][] to be set up
in Kubernetes for the Agent to access it via the Kubernetes REST API.
{{% /admonition %}}
```

File: docs/sources/flow/reference/components/loki.rules.kubernetes.md [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+`use_legacy_routes`      | `bool`     | Whether to use deprecated ruler API endpoints.           | false   | no
+`sync_interval`          | `duration` | Amount of time between reconciliations with Loki.       | "30s"   | no
+`loki_namespace_prefix` | `string`   | Prefix used to differentiate multiple agent deployments. | "agent" | no
+`bearer_token`           | `secret`   | Bearer token to authenticate with.                       |         | no
+`bearer_token_file`      | `string`   | File containing a bearer token to authenticate with.     |         | no
+`proxy_url`              | `string`   | HTTP proxy to proxy requests through.                    |         | no
+`follow_redirects`       | `bool`     | Whether redirects returned by the server should be followed. | `true` | no
+`enable_http2`           | `bool`     | Whether HTTP2 is supported for requests.                 | `true`  | no
+
+ At most one of the following can be provided:
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/loki.rules.kubernetes.md:56]:
```suggestion
 At most, one of the following can be provided:
 
```

File: docs/sources/flow/reference/components/loki.rules.kubernetes.md [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+ - [`authorization` block][authorization].
+ - [`oauth2` block][oauth2].
+
+ [arguments]: #arguments
+
+If no `tenant_id` is provided, the component assumes that the Loki instance at
+`address` is running in single-tenant mode and no `X-Scope-OrgID` header is sent.
+
+The `sync_interval` argument determines how often Loki's ruler API is accessed
+to reload the current state of rules. Interaction with the Kubernetes API works
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/loki.rules.kubernetes.md:69]:
```suggestion
to reload the current state. Interaction with the Kubernetes API works
```

File: docs/sources/flow/reference/components/loki.rules.kubernetes.md [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+differently. Updates are processed as events from the Kubernetes API server
+according to the informer pattern.
+
+The `loki_namespace_prefix` argument can be used to separate the rules managed
+by multiple agent deployments across your infrastructure. It should be set to a
+unique value for each deployment.
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/loki.rules.kubernetes.md:75]:
```suggestion
You can use the `loki_namespace_prefix` argument to separate the rules managed
by multiple agent deployments across your infrastructure. You should set the prefix to a
unique value for each deployment.
```

**Review by clayton-cornell — COMMENTED** [review]
Some suggestions for the docs

**Comment by tpaschalis** [comment]
Hey @EStork09 how's it going? 👋 Would you have the time to look through the docs suggestions and rebase so we can move ahead with this?

File: pkg/loki/client/client.go [hunk 6]
```diff
@@ -0,0 +1,159 @@
+package client
```
[the remark this answers is not in the document]
Comment (EStork09) [reply pkg/loki/client/client.go:1]: I don't disagree, I am just mirroring the mimir client which is also `package client`, arguably these could become a very similar client that is shared, but I am unsure on how the team would like to make this into a reality.

File: docs/sources/flow/reference/components/loki.rules.kubernetes.md [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+* You can specify multiple `loki.rules.kubernetes` components by giving them different labels.
+* [Kubernetes label selectors][] can be used to limit the `Namespace` and
+  `PrometheusRule` resources considered during reconciliation.
+* Compatible with the Ruler APIs of Grafana Loki, Grafana Cloud, and Grafana Enterprise Metrics.
+* Compatible with the `PrometheusRule` CRD from the [prometheus-operator][].
+* This component accesses the Kubernetes REST API from [within a Pod][].
+
+{{% admonition type="note" %}}
+This component requires [Role-based access control (RBAC)][] to be set up
+in Kubernetes for the Agent to access it via the Kubernetes REST API.
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/loki.rules.kubernetes.md:23]:
```suggestion
in Kubernetes for {{< param "PRODUCT_ROOT_NAME" >}} to access it via the Kubernetes REST API.
```

File: docs/sources/flow/reference/components/loki.rules.kubernetes.md [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+
+`loki.rules.kubernetes` supports the following arguments:
+
+Name                     | Type       | Description                                              | Default | Required
+-------------------------|------------|----------------------------------------------------------|---------|---------
+`address`                | `string`   | URL of the Loki ruler.                                  |         | yes
+`tenant_id`              | `string`   | Loki tenant ID.                                         |         | no
+`use_legacy_routes`      | `bool`     | Whether to use deprecated ruler API endpoints.           | false   | no
+`sync_interval`          | `duration` | Amount of time between reconciliations with Loki.       | "30s"   | no
+`loki_namespace_prefix` | `string`   | Prefix used to differentiate multiple agent deployments. | "agent" | no
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/loki.rules.kubernetes.md:49]:
```suggestion
`loki_namespace_prefix` | `string`   | Prefix used to differentiate multiple {{< param "PRODUCT_ROOT_NAME" >}} deployments. | "agent" | no
```

File: docs/sources/flow/reference/components/loki.rules.kubernetes.md [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+If no `tenant_id` is provided, the component assumes that the Loki instance at
+`address` is running in single-tenant mode and no `X-Scope-OrgID` header is sent.
+
+The `sync_interval` argument determines how often Loki's ruler API is accessed
+to reload the current state. Interaction with the Kubernetes API works
+differently. Updates are processed as events from the Kubernetes API server
+according to the informer pattern.
+
+You can use the `loki_namespace_prefix` argument to separate the rules managed
+by multiple agent deployments across your infrastructure. You should set the prefix to a
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/loki.rules.kubernetes.md:74]:
```suggestion
by multiple {{< param "PRODUCT_ROOT_NAME" >}} deployments across your infrastructure. You should set the prefix to a
```

File: docs/sources/flow/reference/components/loki.rules.kubernetes.md [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+    basic_auth {
+        username = "GRAFANA_CLOUD_USER"
+        password = "GRAFANA_CLOUD_API_KEY"
+        // Alternatively, load the password from a file:
+        // password_file = "GRAFANA_CLOUD_API_KEY_PATH"
+    }
+}
+```
+
+The following example is an RBAC configuration for Kubernetes. It authorizes the Agent to query the Kubernetes REST API:
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/loki.rules.kubernetes.md:226]:
```suggestion
The following example is an RBAC configuration for Kubernetes. It authorizes {{< param "PRODUCT_ROOT_NAME" >}} to query the Kubernetes REST API:
```

**Review by clayton-cornell — COMMENTED** [review]
Small cleanup of a few `agent` that slipped through the first review

File: docs/sources/flow/reference/components/loki.rules.kubernetes.md [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+* [Kubernetes label selectors][] can be used to limit the `Namespace` and
+  `PrometheusRule` resources considered during reconciliation.
+* Compatible with the Ruler APIs of Grafana Loki, Grafana Cloud, and Grafana Enterprise Metrics.
+* Compatible with the `PrometheusRule` CRD from the [prometheus-operator][].
+* This component accesses the Kubernetes REST API from [within a Pod][].
+
+{{% admonition type="note" %}}
+This component requires [Role-based access control (RBAC)][] to be set up
+in Kubernetes for {{< param "PRODUCT_ROOT_NAME" >}} to access it via the Kubernetes REST API.
+{{% /admonition %}}
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/loki.rules.kubernetes.md:24]:
```suggestion

Role-based access control (RBAC)]: https://kubernetes.io/docs/reference/access-authn-authz/rbac/
{{% /admonition %}}
```

We need to move the link definition inside the admonition shortcode for the rendering to work correctly.

File: docs/sources/flow/reference/components/loki.rules.kubernetes.md [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+
+[Kubernetes label selectors]: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
+[prometheus-operator]: https://prometheus-operator.dev/
+[within a Pod]: https://kubernetes.io/docs/tasks/run-application/access-api-from-pod/
+[Role-based access control (RBAC)]: https://kubernetes.io/docs/reference/access-authn-authz/rbac/
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/loki.rules.kubernetes.md:29]:
```suggestion
[within a Pod]: https://kubernetes.io/docs/tasks/run-application/access-api-from-pod/
```

**Review by clayton-cornell — COMMENTED** [review]
Additional docs fixes

**Comment by wildum** [comment]
To avoid duplicating so much code it should be possible to put some in the common kubernetes pkg. The client part could also use a common pkg. But we can do this via a follow-up PR and focus on getting this one merged

**Comment by wildum** [comment]
I tested it and it works: I created a local k8s cluster with the agent using the helm chart and the prometheus operator. Then I created the following rule:
```
apiVersion: monitoring.coreos.com/v1
kind: PrometheusRule
metadata:
  creationTimestamp: null
  labels:
    prometheus: example
    role: alert-rules
  name: prometheus-example-rules
spec:
  groups:
  - name: ./example.rules
    rules:
    - alert: ExampleAlert
      expr: rate({job="test", container="my-app"} |= "error" [5m]) > 5
```
and with this component:
```
loki.rules.kubernetes "hello" {
          address = "loki-url"
          basic_auth {
              username = "username"
              password = "token"
          }
      }
```
It successfully pushed the alert to my grafana cloud stack:
Screenshot 2024-02-27 at 18 24 51

File: docs/sources/flow/reference/components/loki.rules.kubernetes.md [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+* Compatible with the Ruler APIs of Grafana Loki, Grafana Cloud, and Grafana Enterprise Metrics.
+* Compatible with the `PrometheusRule` CRD from the [prometheus-operator][].
+* This component accesses the Kubernetes REST API from [within a Pod][].
+
+{{% admonition type="note" %}}
+This component requires [Role-based access control (RBAC)][] to be set up
+in Kubernetes for {{< param "PRODUCT_ROOT_NAME" >}} to access it via the Kubernetes REST API.
+
+Role-based access control (RBAC)]: https://kubernetes.io/docs/reference/access-authn-authz/rbac/
+{{% /admonition %}}
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/loki.rules.kubernetes.md:26]:
```suggestion
{{< admonition type="note" >}}
This component requires [Role-based access control (RBAC)][] to be set up
in Kubernetes for {{< param "PRODUCT_ROOT_NAME" >}} to access it via the Kubernetes REST API.

Role-based access control (RBAC)]: https://kubernetes.io/docs/reference/access-authn-authz/rbac/
{{< /admonition >}}
```

Minor shortcode syntax update

**Review by tpaschalis — COMMENTED** [review]
LGTM, but I haven't personally tested this so I'll leave the final word for @wildum 

In any case, great work @EStork09 !

File: component/loki/rules/kubernetes/rules.go [hunk 14]
```diff
@@ -0,0 +1,335 @@
+package rules
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
```
Comment (wildum) [inline component/loki/rules/kubernetes/rules.go:10]: should be "github.com/grafana/agent/pkg/flow/logging/level"

File: component/loki/rules/kubernetes/events.go [hunk 15]
```diff
@@ -0,0 +1,270 @@
+package rules
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
```
Comment (wildum) [inline component/loki/rules/kubernetes/events.go:10]: should be "github.com/grafana/agent/pkg/flow/logging/level"

File: component/loki/rules/kubernetes/rules.go [hunk 16]
```diff
@@ -0,0 +1,335 @@
...
+	err = c.init()
+	if err != nil {
+		return nil, fmt.Errorf("initializing component failed: %w", err)
+	}
+
+	return c, nil
+}
+
+func (c *Component) Run(ctx context.Context) error {
+	err := c.startup(ctx)
```
[the remark this answers is not in the document]
Comment (wildum) [reply component/loki/rules/kubernetes/rules.go:156]: -> this is a recent change in mimir, I will bring it also to loki

[the remark this answers is not in the document]
Comment (wildum) [reply component/loki/rules/kubernetes/rules.go:156]: see https://github.com/grafana/agent/pull/5906 for more info

**Review by wildum — APPROVED** [review]
Made some little changes and updated with fresh main. Retested after the changes and everything works.

**Comment by wildum** [comment]
Thansk again @EStork09 and sorry it took so long!

**Pull request merged**